### PR TITLE
cmd: Rename `jj touchup` to `jj diffedit`, consolidate it with `jj restore -i`

### PR DESCRIPTION
**Pull request by ilyagr**
This allows touching up a commit while seeing the differences between it and
another arbitrary commit.

For example, you can view diff between arbitrary commits in a diff editor.  You
can also do `jj update abc; jj touchup --diff-from @--` as a safer alternative
to `jj edit abc` if there are many commits on top of `abc` and rebasing them
constantly causes data races.

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [x] I have added myself to the contributors in `CHANGELOG.md` (optional)

### CODE REVIEW

**Comment by martinvonz** [comment]
Hmm, does that mean that `jj touch --diff-from foo -r bar` is the same as `jj restore -i --from foo --to bar`?

**Comment by ilyagr** [comment]
I guess so. I didn't think of that, because in my mind these are in different places: `jj touchup` is what I do when I want to see a commit in a difftool and potentially edit it. `jj restore` if what I do to, well, pull something in from another commit. Something feels off about using `restore` to edit a commit in unrelated ways, though I'm sure it's possible to get used to.

In fact, I guess that except for merge commits, `touchup` is a special case of `restore -i`. For merge commits, you can simulate `touchup` with a `merge` and a `restore -i`.

I think there's a place for both, but it's up to you.

**Comment by martinvonz** [comment]
> I guess so, @martinvonz . I didn't think of that, because in my mind these are in different places: `jj touchup` is what I do when I want to see a commit in a difftool and potentially edit it. `jj restore` if what I do to, well, pull something in from another commit. Something feels off about using `restore` to edit a commit in unrelated ways, though I'm sure it's possible to get used to.

That all makes sense.

I hadn't actually thought about using `touchup` just for viewing a diff until I saw this PR. I wonder if we should rename the command to better fit that usecase. It used to be called `edit`, but that was before our current `edit` command existed. I wonder if `edit-diff` would be better than `touchup`. It seems like that matches pretty well what it does. Is it too similar to `edit`?

> 
> In fact, I guess that except for merge commits, `touchup` is a special case of `restore -i`. For merge commits, you can simulate `touchup` with a `merge` and a `restore -i`.

Yes, and `squash` and `unsquash` are special cases of `move`.

> I think there's a place for both, but it's up to you.

Yeah, that's probably fine.

File: src/commands.rs [hunk 1]
```diff
@@ -566,6 +566,13 @@ struct TouchupArgs {
     /// The revision to touch up
     #[arg(long, short, default_value = "@")]
     revision: RevisionArg,
+    /// Revision to show on the left side of the diff
+    ///
+    /// Instead of showing the changes between the revision being edited and
+    /// its parents, show the diff between the revision being edited and the
+    /// revision specified here.
+    #[arg(long, value_name = "REVISION")]
+    diff_from: Option<RevisionArg>,
```
Comment (martinvonz) [inline src/commands.rs:575]: We should probably replace the arguments by `--from` and `--to` and match `jj diff`'s behavior. This command is very similar to `jj diff`, and can even be used for viewing a diff in a GUI, as you pointed out. See `DiffArgs` for how we tell Clap which arguments are incompatible. Agree?

Comment (ilyagr) [reply src/commands.rs:575]: I don't know if it makes sense to have separate `-r` and `--to`. With `--to`, it's less clear that the `--to` commit will be modified if the user makes any edits to the diff. Perhaps it would make sense to edit the *left* side of the commit if `--to` is specified, but not `--from`, but that's both complicated to implement and potentially confusing.

**Comment by ilyagr** [comment]
>  I wonder if `edit-diff` would be better than `touchup`

I would suggest `diffedit` (my preference) or `diff-edit`. `git` and `hg` call it `difftool`, I believe, so I think it's nice to have a similar name. This would go well with making the arguments similar to `diff` (though I have some worries about that, as I said in the other comment).

**Comment by martinvonz** [comment]
> > I wonder if `edit-diff` would be better than `touchup`
> 
> I would suggest `diffedit` (my preference) or `diff-edit`. `git` and `hg` call it `difftool`, I believe, so I think it's nice to have a similar name.

Good point to compare with `git` and `hg`. You're right that `git difftool`, but it's `hg extdiff`. `diffedit` sounds good to me. Let's see if anyone else has any thoughts.

File: src/commands.rs (hunk 1 above)
Comment (martinvonz) [reply src/commands.rs:575]: I kind of see what you mean about `jj diffedit --to foo`, but I also feel like I would expect that to edit the `--to` side rather than the working copy. Btw, I have tried to avoid having one argument be called the generic `-r/--revision` when there are revisions that play different roles involved (like `--diff-from` and `-r`).

I also wonder if we should give `jj restore` an `-r` argument and make these commands aliases. It seems like they will do exactly the same thing, just described differently. Thanks for making me see the similarity between `touchup` and `restore`.

**Comment by yuja** [comment]
`diffedit` sounds good to me. Regarding command options, I second `--from`/`--to` than `--from`/`-r`. If `jj` had `hg extdiff`/`git difftool` functionality, I think it should take the same set of options as `jj diff`, so does `jj diffedit`.

File: src/commands.rs (hunk 1 above)
Comment (ilyagr) [reply src/commands.rs:575]: I'll resolve this conversation and move it to the main body.

**Comment by ilyagr** [comment]
(Replying both to the main conversation and the now-resolved conversation)

I think we're converging on the following:

* `jj diffedit abc` or `jj diffedit -r abc` work as `touchup` does now.
* `-r` cannot be used together with `--form`/`--to`.
* `jj diffedit --to abc` is the same as `jj diffedit --from @ --to abc` and edits `abc`.
* `jj diffedit --from abc` is the same as `jj diffedit --from abc --to @` and edits `@`.

This sounds good to me, but let me know if we're not on the same page.

> I also wonder if we should give jj restore an -r argument and make these commands aliases. 

My first instinct is to remove the `-i` argument to `restore`, pointing out in the docs to use `diffedit` for that instead. I like that `jj restore file` and `jj diffedit abc` do the intuitive thing without having to insert an argument. Moreover, these would now be very similar commands, but one is reserved for command-line use and the other for using an external tool. We could mention the other command in each command's help text.

**Comment by ilyagr** [comment]
I haven't reviewed this myself super-carefully, but I think it's ready for you to look at it.

Let me know if there are any uses of `jj restore -i` in the docs I need to fix.

File: docs/tutorial.md [hunk 2]
```diff
@@ -390,10 +390,10 @@ When Meld starts, edit the right side by e.g. adding something to the first
...
 Unlike `jj squash -i`, which left the content state of the commit unchanged,
-`jj touchup` (typically) results in a different state, which means that
+`jj diffedit` (typically) results in a different state, which means that
 descendant commits may have conflicts.
 
-Other commands for rewriting contents of existing commits are `jj restore -i`,
-`jj split`, `jj unsquash -i`. Now that you've seen how `jj squash -i` and
-`jj touchup` work, you can hopefully figure out how those work (with the help of
-the instructions in the diff).
+Other commands for rewriting contents of existing commits are `jj split` and
```
Comment (ilyagr) [inline docs/tutorial.md:396]: Should `jj move` be mentioned here?

File: src/commands.rs [hunk 3]
```diff
@@ -530,27 +533,40 @@ struct RestoreArgs {
...
-/// close the editor, the revision will be updated. Descendants will be rebased
-/// on top as usual, which may result in conflicts. See `jj squash -i` or `jj
-/// unsquash -i` if you instead want to move changes into or out of the parent
-/// revision.
+/// close the editor, the revision specified with `-r` or `--to` will be
+/// updated. Descendants will be rebased on top as usual, which may result in
+/// conflicts.
+///
+/// See `jj restore` if you want to move entire files from one revision to
+/// another. See `jj squash -i` or `jj unsquash -i` if you instead want to move
```
Comment (ilyagr) [inline src/commands.rs:555]: Should these perhaps be removed as well and replaced with `jj diffedit --to @-` and `jj diffedit --from @-`? (May or may not be part of this PR)

Comment (ilyagr) [reply src/commands.rs:555]: I guess maybe not. Those are instances of `jj move` , not  `jj diffedit`.

File: docs/tutorial.md (hunk 2 above)
Comment (martinvonz) [reply docs/tutorial.md:396]: Yes, that's one of the more important ones, so that would make sense. Thanks.

**Review by martinvonz — APPROVED** [review]
Looks great. Thank you!
